internal/runtime/catalog/test/catalog_test.go

```diff
@@ -0,0 +1,282 @@
+
+		// Test NewRequest
+		request, err := c.NewRequest(hookGVH)
+		g.Expect(err).ToNot(HaveOccurred())
+
+		// Test NewResponse
+		response, err := c.NewResponse(hookGVH)
+		g.Expect(err).ToNot(HaveOccurred())
+
+		// Test ValidateRequest/ValidateResponse
+		g.Expect(c.ValidateRequest(hookGVH, request)).To(Succeed())
+		g.Expect(c.ValidateResponse(hookGVH, response)).To(Succeed())
+	}
+
+	verify(v1alpha1.FakeHook, v1alpha1.GroupVersion)
+	verify(v1alpha2.FakeHook, v1alpha2.GroupVersion)
+}
+
+func TestValidateRequest(t *testing.T) {
+	v1alpha1Hook, err := c.GroupVersionHook(v1alpha1.FakeHook)
+	if err != nil {
+		panic("failed to get GVH of hook")
+	}
+	v1alpha1HookRequest, err := c.NewRequest(v1alpha1Hook)
+	if err != nil {
+		panic("failed to create request for hook")
+	}
+
+	v1alpha2Hook, err := c.GroupVersionHook(v1alpha2.FakeHook)
+	if err != nil {
+		panic("failed to get GVH of hook")
+	}
+
+	tests := []struct {
+		name      string
+		hook      catalog.GroupVersionHook
+		request   runtime.Object
+		wantError bool
+	}{
+		{
+			name:      "should succeed when hook and request match",
+			hook:      v1alpha1Hook,
+			request:   v1alpha1HookRequest,
+			wantError: false,
+		},
+		{
+			name:      "should error when hook and request do not match",
+			hook:      v1alpha2Hook,
+			request:   v1alpha1HookRequest,
+			wantError: true,
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			g := NewWithT(t)
+			err := c.ValidateRequest(tt.hook, tt.request)
+			if tt.wantError {
+				g.Expect(err).To(HaveOccurred())
+			} else {
+				g.Expect(err).NotTo(HaveOccurred())
+			}
+		})
+	}
+}
+
+func TestValidateResponse(t *testing.T) {
+	v1alpha1Hook, err := c.GroupVersionHook(v1alpha1.FakeHook)
+	if err != nil {
+		panic("failed to get GVH of hook")
+	}
+	v1alpha1HookResponse, err := c.NewResponse(v1alpha1Hook)
+	if err != nil {
+		panic("failed to create request for hook")
+	}
+
+	v1alpha2Hook, err := c.GroupVersionHook(v1alpha2.FakeHook)
+	if err != nil {
+		panic("failed to get GVH of hook")
+	}
+
+	tests := []struct {
+		name      string
+		hook      catalog.GroupVersionHook
+		response  runtime.Object
+		wantError bool
+	}{
+		{
+			name:      "should succeed when hook and response match",
+			hook:      v1alpha1Hook,
+			response:  v1alpha1HookResponse,
+			wantError: false,
+		},
+		{
+			name:      "should error when hook and response do not match",
+			hook:      v1alpha2Hook,
+			response:  v1alpha1HookResponse,
+			wantError: true,
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			g := NewWithT(t)
+			err := c.ValidateResponse(tt.hook, tt.response)
+			if tt.wantError {
+				g.Expect(err).To(HaveOccurred())
+			} else {
+				g.Expect(err).NotTo(HaveOccurred())
+			}
+		})
+	}
+}
+
+type GoodRequest struct {
+	metav1.TypeMeta `json:",inline"`
+
+	First string `json:"first"`
+}
+
+func (in *GoodRequest) DeepCopyObject() runtime.Object {
```

Comment:
Q: What do you think about generating the deepcopy stuff for the fake types the same way that we generate conversion? It's probably simpler to adjust the Makefile a bit vs. adding the funcs over time with panics :)